File: clerks/Turtle/Turtle.java
import static java.lang.StringTemplate.STR;

class Turtle implements Clerk {
    final String ID;
    LiveView view;
    final int width, height;

    Turtle(LiveView view, int width, int height) {
        this.view = view;
        this.width  = Math.max(1, Math.abs(width));  // width is at least of size 1
        this.height = Math.max(1, Math.abs(height)); // height is at least of size 1
        ID = Clerk.getHashID(this);
        Clerk.load(view, "clerks/Turtle/turtle.js");
        Clerk.write(view, STR."""
            <canvas id="turtleCanvas\{ID}" width="\{this.width}" height="\{this.height}" style="border:1px solid #000;">
            </canvas>
            """);
        Clerk.script(view, STR."const turtle\{ID} = new Turtle(document.getElementById('turtleCanvas\{ID}'));");
    }

    Turtle(LiveView view) { this(view, 500, 500); }
    Turtle(int width, int height) { this(Clerk.view(), width, height); }
    Turtle() { this(Clerk.view()); }

    Turtle penDown() {
        Clerk.call(view, STR."turtle\{ID}.penDown();");
        return this;
    }

    Turtle penUp() {
        Clerk.call(view, STR."turtle\{ID}.penUp();");
        return this;
    }

    Turtle forward(double distance) {
        Clerk.call(view, STR."turtle\{ID}.forward(\{distance});");
        return this;
    }

    Turtle backward(double distance) {
        Clerk.call(view, STR."turtle\{ID}.backward(\{distance});");
        return this;
    }

    Turtle left(double degrees) {
        Clerk.call(view, STR."turtle\{ID}.left(\{degrees});");
        return this;
    }

    Turtle right(double degrees) {
        Clerk.call(view, STR."turtle\{ID}.right(\{degrees});");
        return this;
    }

    Turtle color(String color) {
        Clerk.call(view, STR."turtle\{ID}.color('\{color}');");
        return this;
    }

    Turtle lineWidth(double width) {
        Clerk.call(view, STR."turtle\{ID}.lineWidth('\{width}')");
        return this;
    }
}

File: clerks/Turtle/turtle.js
class Turtle {
    constructor(canvas) {
        this.canvas = canvas;
        this.ctx = canvas.getContext('2d');
        this.reset();
    }

    reset() {
        this.ctx.reset();
        this.x = this.canvas.width / 2;
        this.y = this.canvas.height / 2;
        this.angle = 0;
        this.penDown();
        this.color("black");
    }

    penDown() {
        this.isPenDown = true;
    }

    penUp() {
        this.isPenDown = false;
    }

    forward(distance) {
        const radians = (this.angle * Math.PI) / 180;
        const newX = this.x + distance * Math.cos(radians);
        const newY = this.y + distance * Math.sin(radians);

        if (this.isPenDown) {
            this.ctx.beginPath();
            this.ctx.moveTo(this.x, this.y);
            this.ctx.lineTo(newX, newY);
            this.ctx.stroke();
        }

        this.x = newX;
        this.y = newY;
    }

    backward(distance) {
        this.forward(-distance);
    }

    right(degrees) {
        this.angle += degrees;
    }

    left(degrees) {
        this.angle -= degrees;
    }

    color(color) {
        this.ctx.strokeStyle = color;
    }

    lineWidth(width) {
        this.ctx.lineWidth = width;
    }

    text(text, font = '10px sans-serif', align = 'center') {
        const radians = (this.angle * Math.PI) / 180 + Math.PI / 2.0;
        this.ctx.save();
        this.ctx.translate(this.x, this.y);
        this.ctx.rotate(radians);
        this.ctx.font = font;
        this.ctx.fillStyle = this.ctx.strokeStyle;
        this.ctx.textAlign = align;
        this.ctx.fillText(text, 0, 0);
        this.ctx.restore();
    }
    moveTo(x, y) {
        this.x = x;
        this.y = y;
    }

    lineTo(x, y) {
        const originalPenState = this.isPenDown;
        this.isPenDown = true;

        this.ctx.beginPath();
        this.ctx.moveTo(this.x, this.y);
        this.ctx.lineTo(x, y);
        this.ctx.stroke();

        this.x = x;
        this.y = y;

        this.isPenDown = originalPenState;
    }
}
